### ISSUE TITLE
Undeploy/redeploy fails to recreate disk.0 symlink

### ISSUE THREAD

**Issue by mrschyte** [OpenNebula/addon-linstor_un#4]
Undeploying and redeploying a VM results in failure. The /var/lib/one/datastores/XXX/YYY/disk.0 symlink is not recreated when redeploying the machine.

The failure can be resolved by manually recreating the symlink (e.g. "ln -s /dev/drbd/by-res/one-vm-YYY-disk-0/0 disk.0" and  doing a recover/success operation.

linstor_un should implement a custom vmm/deploy script which recreates the symlink automatically to prevent this issue.

**Comment by kvaps** [comment]
Hi, when you run undeploy/deploy operations OpenNebula calls [`/var/lib/one/remotes/tm/linstor_un/mv`](https://github.com/OpenNebula/addon-linstor_un/blob/master/tm/linstor_un/mv) script with parameters for the VM.
Actually it should fix the symlinks.

Could you check the VM log and provide more details about your case or exact steps to reproduce?

Thank You!

**Comment by mrschyte** [comment]
Thanks for the quick response. I have two nodes set up (nebula01, nebula02) with the frontend running in an active/passive HA solution using pacemaker (both nodes are added as compute nodes).

The strange thing is that the redeploy only fails if I'm trying to move a machine from the secondary node to the primary node which is running the frontend. If I undeploy a machine on the frontend I can redeploy it to the secondary node but not to itself. Migration (both live & offline) work fine, so I'm not sure why undeploy/deploy doesn't.

I've attached the log file for the VM, but I don't see anything that would reveal why it's not recreating the symlink.
```
Wed Oct 23 11:19:05 2019 [Z0][VMM][I]: Command execution fail: cat << EOT | /var/tmp/one/vmm/kvm/deploy '/var/lib/one//datastores/106/183/deployment.2' 'nebula01' 183 nebula01
Wed Oct 23 11:19:05 2019 [Z0][VMM][I]: error: Failed to create domain from /var/lib/one//datastores/106/183/deployment.2
Wed Oct 23 11:19:05 2019 [Z0][VMM][I]: error: Cannot access storage file '/var/lib/one//datastores/106/183/disk.0': No such file or directory
Wed Oct 23 11:19:05 2019 [Z0][VMM][E]: Could not create domain from /var/lib/one//datastores/106/183/deployment.2
Wed Oct 23 11:19:05 2019 [Z0][VMM][I]: ExitCode: 255
```

https://gist.github.com/mrschyte/b39e9f680abbca90fc08bf70ae8829ac

**Comment by kvaps** [comment]
Thanks for the info I checked the provided logs.

If I understood it well problem occurs only when deploying VM on the current frontend node.
According the logs, script [`/var/lib/one/remotes/tm/linstor_un/mv`](https://github.com/OpenNebula/addon-linstor_un/blob/master/tm/linstor_un/mv) finishes without any errors or OpenNebula just not calling it in this case for some reason.

The only reason why script might finish without errors and do no create any symlinks is when VM in `EPILOG_STOP`, `EPILOG_UNDEPLOY` state, that's probably not your case.
https://github.com/OpenNebula/addon-linstor_un/blob/07e64eadec464d436299f1ef14b091114d18419f/tm/linstor_un/mv#L153-L156

So looks like it is upstream problem, could you repeat your steps, but with one change to confirm that.

* Undeploy the VM
* Update `/var/lib/one/remotes/tm/linstor_un/mv` script, add `exit 1` into second line
* Try to deploy your VM on current Frontend server.
* Check the logs, if OpenNebula calls the `linstor_un/mv` script, it should throw an error, if not, you will see the error about `kvm/deploy` script as you sent above.

**Comment by mrschyte** [comment]
It seems that the `mv` script is called as adding `exit 1` results in the following error:
```
Wed Oct 23 12:44:53 2019 [Z0][VM][I]: New state is ACTIVE
Wed Oct 23 12:44:53 2019 [Z0][VM][I]: New LCM state is PROLOG_UNDEPLOY
Wed Oct 23 12:44:53 2019 [Z0][TM][I]: Command execution failed (exit code: 1): /var/lib/one/remotes/tm/linstor_un
/mv nebula02:/var/lib/one//datastores/106/189/disk.0 nebula02:/var/lib/one//datastores/106/189/disk.0 189 107
Wed Oct 23 12:44:53 2019 [Z0][TM][E]: Error executing image transfer script
Wed Oct 23 12:44:53 2019 [Z0][VM][I]: New LCM state is PROLOG_UNDEPLOY_FAILURE
Wed Oct 23 12:46:42 2019 [Z0][VM][I]: New LCM state is PROLOG_UNDEPLOY
Wed Oct 23 12:46:43 2019 [Z0][VM][I]: New LCM state is BOOT_UNDEPLOY
```
The problem seems to be that OpenNebula is calling `mv` with both the source and the destination set to the current frontend node (nebula02) and the `mv` script will happily remove the symlink after creating it.

Adding the following check at the end fixes the issue:
```
if [[ "$SRC_HOST" != "$DST_HOST" ]] || [[ "$SRC_PATH" != "$DST_PATH" ]]; then
  ssh_exec_and_log_no_error "$SRC_HOST" \
      "rm -rf $SRC_PATH" \
      "Error removing VM files on System DS: $SRC_PATH on $SRC_HOST"
fi
```

**Comment by kvaps** [comment]
OK, now I see, but problem is that this behavior is unexpected.
When VM on `PROLOG_UNDEPLOY` state it should run `exit 0` before:

https://github.com/OpenNebula/addon-linstor_un/blob/07e64eadec464d436299f1ef14b091114d18419f/tm/linstor_un/mv#L186-L189

Could you check the `LCM_STATE` of the VM, on time when it is happening?
Just add a line before this snippet please:
```bash
echo "$LCM_STATE" >> /tmp/123
```

and check `/tmp/123` file

**Comment by mrschyte** [comment]
Something unsets `LCM_STATE`. I've printed it right after it's set and it contains the correct values. However, if I print it at the end, before the `PROLOG_RESUME + PROLOG_UNDEPLOY` block, it's empty.

**Comment by mrschyte** [comment]
Hmm. Actually it's not unset, it's just not set. The first if block (Activate the disk in target host) reads the `LCM_STATE` and that's skipped on `PROLOG_UNDEPLOY`. Moving the following block to the beginning after the `DST_DIR` variable works.
```
unset i j XPATH_ELEMENTS

while IFS= read -r -d '' element; do
XPATH_ELEMENTS[i++]="$element"
done < <(onevm show -x $VMID | $XPATH \
                /VM/LCM_STATE)

LCM_STATE="${XPATH_ELEMENTS[j++]}"
```

**Comment by kvaps** [comment]
OK I see, thank you for PR, but I want to know why actually is it happening.
Looks like OpenNebula calling:

```
/var/lib/one/remotes/tm/linstor_un
/mv nebula02:/var/lib/one//datastores/106/189/disk.0 nebula02:/var/lib/one//datastores/106/189/disk.0 189 107
```
which is working fine, but after that it is calling:
```
/var/lib/one/remotes/tm/linstor_un
/mv nebula02:/var/lib/one//datastores/106/189/ nebula02:/var/lib/one//datastores/106/189/ 189 107
```
Which is removes the symlink.

Excuse me for my meticulousness, but could you please recover original `mv` and add this block into second line:

```bash
if [ "$DEBUG" != "1" ]; then
  echo "[`date`] $0 $@" >> /tmp/123
  DEBUG=1 bash -x "$0" "$@" >>/tmp/123 2>&1 
  exit $?
fi
```
then try again and provide me `/tmp/123` file:

I'll be very grateful to you! 🙂

**Comment by mrschyte** [comment]
No problem. Here you go. I've split the log into two files, the first is when undeploying the machine and the second is when redeploying it (which fails).
https://gist.github.com/mrschyte/95f60d7c30616f926c22cae1948a0ec6

**Comment by kvaps** [comment]
My guesses were true, thanks for cooperation!
Your PR merged into `v1.2.5`
